File: mbw/src/main/java/com/mycelium/bequant/kyc/verticalStepper/VerticalStepperItemView.java
package com.mycelium.bequant.kyc.verticalStepper;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Typeface;
import android.os.Build;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.res.ResourcesCompat;

import com.mycelium.wallet.R;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

public class VerticalStepperItemView extends FrameLayout {
    public static int STATE_INACTIVE = 0;

    public static int STATE_ACTIVE = 1;

    public static int STATE_COMPLETE = 2;

    private boolean showConnectorLine = true;

    private boolean editable = false;

    private VerticalStepperItemCircleView circle;

    private int number;

    private LinearLayout wrapper;

    private TextView title;


    private ConnectorLineDrawer connector;

    private int state = STATE_INACTIVE;

    public VerticalStepperItemView(Context context) {
        super(context);
        initialize(context);
    }

    public VerticalStepperItemView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initialize(context);
    }

    public VerticalStepperItemView(
            Context context,
            AttributeSet attrs,
            int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize(context);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public VerticalStepperItemView(
            Context context,
            AttributeSet attrs,
            int defStyleAttr,
            int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initialize(context);
    }

    private void initialize(Context context) {
        setWillNotDraw(false);
        setClipChildren(false);
        setClipToPadding(false);

        int padding = (int) Util.dpToPx(context, 8);
        setPadding(padding, padding, padding, 0);

        LayoutInflater.from(context).inflate(
                R.layout.vertical_stepper_view_item,
                this,
                true);

        circle = (VerticalStepperItemCircleView) findViewById(R.id.vertical_stepper_view_item_circle);
        wrapper = (LinearLayout) findViewById(R.id.vertical_stepper_view_item_wrapper);
        title = (TextView) findViewById(R.id.vertical_stepper_view_item_title);
        connector = new ConnectorLineDrawer(context);
    }

    public boolean getShowConnectorLine() {
        return showConnectorLine;
    }

    public void setShowConnectorLine(boolean show) {
        showConnectorLine = show;
        setMarginBottom(state == STATE_ACTIVE);
    }

    public boolean isEditable() {
        return editable;
    }

    public void setEditable(boolean editable) {
        this.editable = editable;

        if (state == STATE_COMPLETE)
            if (isEditable())
                circle.setIconEdit();
            else
                circle.setIconCheck();
    }

    public void setCircleNumber(int number) {
        this.number = number;

        if (state != STATE_COMPLETE)
            circle.setNumber(number);
    }

    public void setTitle(CharSequence title) {
        this.title.setText(title);
    }

    public int getState() {
        return state;
    }

    public void setState(int state) {
        this.state = state;

        if (state == STATE_INACTIVE)
            setStateInactive();
        else if (state == STATE_ACTIVE)
            setStateActive();
        else
            setStateComplete();
    }

    private void setStateInactive() {
        circle.setIconEdit();
        setMarginBottom(false);
        circle.setNumber(number);
        circle.setBackgroundInactive();
        title.setTextColor(ResourcesCompat.getColor(
                getResources(),
                R.color.bequant_gray_6,
                null));
        title.setTypeface(title.getTypeface(), Typeface.BOLD);
    }

    private void setStateActive() {
        circle.setIconEdit();
        setMarginBottom(true);
        circle.setNumber(number);
        circle.setBackgroundActive();
        title.setTextColor(ResourcesCompat.getColor(
                getResources(),
                R.color.bequant_gray_6,
                null));
        title.setTypeface(title.getTypeface(), Typeface.BOLD);
    }

    private void setStateComplete() {
        setMarginBottom(false);
        circle.setBackgroundActive();

        if (isEditable())
            circle.setIconEdit();
        else
            circle.setIconCheck();

        title.setTextColor(ResourcesCompat.getColor(
                getResources(),
                R.color.white,
                null));
        title.setTypeface(title.getTypeface(), Typeface.BOLD);

    }

    private void setMarginBottom(boolean active) {
        MarginLayoutParams params = (MarginLayoutParams) wrapper
                .getLayoutParams();

        if (!getShowConnectorLine())
            params.bottomMargin = 0;
        else if (active)
            params.bottomMargin = (int) Util.dpToPx(getContext(), 48);
        else
            params.bottomMargin = (int) Util.dpToPx(getContext(), 40);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (showConnectorLine)
            connector.draw(canvas);
    }

    @Override
    protected void onSizeChanged(
            int width,
            int height,
            int oldWidth,
            int oldHeight) {
        super.onSizeChanged(width, height, oldWidth, oldHeight);

        connector.adjust(getContext(), width, height);
    }
}

File: mbw/src/main/java/com/mycelium/bequant/kyc/verticalStepper/VerticalStepperItemView.kt
package com.mycelium.bequant.kyc.verticalStepper

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import androidx.annotation.ColorRes
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.res.ResourcesCompat
import com.mycelium.bequant.kyc.steps.adapter.StepState
import com.mycelium.wallet.R
import com.mycelium.wallet.databinding.VerticalStepperViewItemBinding

class VerticalStepperItemView : ConstraintLayout {
    var showConnectorLine = true
    var number = 0
    var title = ""
    var state = StepState.FUTURE

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr)

    private val binding =
        VerticalStepperViewItemBinding.inflate(LayoutInflater.from(context), this)

    fun update() {
        binding.verticalStepperViewItemTitle?.text = title
        binding.verticalStepperViewItemCircle?.setNumber(number)
        when (state) {
            StepState.COMPLETE -> {
                binding.verticalStepperViewItemCircle?.setIconCheck()
                binding.verticalStepperViewItemCircle?.setComplete()
                setTitleColor(R.color.white)
                if (showConnectorLine) {
                    setLineColor(R.color.bequant_green)
                }
            }
            StepState.COMPLETE_EDITABLE -> {
                binding.verticalStepperViewItemCircle?.setIconEdit()
                binding.verticalStepperViewItemCircle?.setActive()
                setTitleColor(R.color.white)
                if (showConnectorLine) {
                    setLineColor(R.color.bequant_yellow)
                }
            }
            StepState.CURRENT -> {
                binding.verticalStepperViewItemCircle?.setActive()
                setTitleColor(R.color.white)
                if (showConnectorLine) {
                    setLineColor(R.color.bequant_yellow)
                }
            }
            StepState.FUTURE -> {
                binding.verticalStepperViewItemCircle?.setInactive()
                setTitleColor(R.color.bequant_gray_6)
                if (showConnectorLine) {
                    setLineColor(R.color.bequant_gray_6)
                }
            }
            StepState.ERROR -> {
                binding.verticalStepperViewItemCircle?.setError()
                binding.verticalStepperViewItemCircle?.setIconError()
                setTitleColor(R.color.white)
                if (showConnectorLine) {
                    setLineColor(R.color.bequant_red)
                }
            }
        }
    }

    private fun setTitleColor(@ColorRes color: Int) {
        binding.verticalStepperViewItemTitle?.setTextColor(ResourcesCompat.getColor(resources, color, null))
    }

    private fun setLineColor(@ColorRes color: Int) {
        binding.line?.setBackgroundColor(ResourcesCompat.getColor(resources, color, null))
    }
}